### PR TITLE
DT-881 nested arrays

### PR DESCRIPTION
**Pull request by mikehalmamoj**
The bad JSON that previously broke the UI was found - it was fine but contained some nested arrays which broke the assumption that all messages contained simple string -> string key value pairs.  We now handle any type of value.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt [hunk 1]
```diff
@@ -14,8 +14,13 @@ class OffenderEventStore(@Value("\${model.cacheSize}") private val cacheSize: In
     fun fromJson(json: String): MutableMap<String, String> {
       return try {
         Gson().fromJson(json, object : TypeToken<MutableMap<String, String>>() {}.type)
-      } catch(e: Exception) {
-        mutableMapOf("BadMessage" to json, "CausedException" to e.toString())
+      } catch (e1: Exception) {
+        try {
+          val anyMap: Map<String, Any> = Gson().fromJson(json, object : TypeToken<Map<String, Any>>() {}.type)
+          anyMap.entries.associate { it.key to it.value.toString() }.toMutableMap()
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt:20]: If the values in the JSON can be deserialized as type Any, we convert them to string and send them back.

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt [hunk 2]
```diff
@@ -14,8 +14,13 @@ class OffenderEventStore(@Value("\${model.cacheSize}") private val cacheSize: In
...
+        } catch (e2: Exception) {
+          mutableMapOf("BadMessage" to json, "CausedException" to e1.toString())
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt:22]: If the JSON genuinely cannot be parsed, we send the error back.

File: src/test/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStoreTest.kt [hunk 3]
```diff
@@ -83,5 +83,19 @@ class OffenderEventStoreTest {
...
+
+    val stored = offenderEventStore.getPageOfMessages(listOf(), listOf(), "", 1)
+    assertThat(stored.get(0).messageDetails["offenderIdDisplay"]).isEqualTo("G0373GG")
+    assertThat(stored.get(0).messageDetails["offenders"]).contains("1025558")
+  }
+
+  private fun aMessage() = Message("{\"ANY_KEY\": \"ANY_MESSAGE\"}", "ANY_MESSAGE_ID", MessageAttributes(EventType("ANY_EVENT_TYPE")))
+
+  private fun aNestedArrayMessage() = Message(
+      "{\"offenderIdDisplay\":\"G0373GG\",\"offenders\":[{\"offenderId\":1025558,\"bookings\":[{\"offenderBookId\":12678}]}]}",
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStoreTest.kt:98]: An example that previously broke the UI.

Comment (brightonsbox) [reply src/test/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStoreTest.kt:98]: Sorry, my fault!!

Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStoreTest.kt:98]: Perfectly valid JSON means you no-claims bonus is fully intact :)

**Pull request merged**